### PR TITLE
Add lint against ambiguous wide pointer comparisons

### PR DESCRIPTION
**Pull request by Urgau**
This PR is the resolution of https://github.com/rust-lang/rust/issues/106447 decided in https://github.com/rust-lang/rust/issues/117717 by T-lang.

## `ambiguous_wide_pointer_comparisons`

*warn-by-default*

The `ambiguous_wide_pointer_comparisons` lint checks comparison of `*const/*mut ?Sized` as the operands.

### Example

```rust
let ab = (A, B);
let a = &ab.0 as *const dyn T;
let b = &ab.1 as *const dyn T;

let _ = a == b;
```

### Explanation

The comparison includes metadata which may not be expected.

-------

This PR also drops `clippy::vtable_address_comparisons` which is superseded by this one.

~~One thing: is the current naming right? `invalid` seems a bit too much.~~

Fixes https://github.com/rust-lang/rust/issues/117717

### CODE REVIEW

**Comment by asquared31415** [comment]
#117717 says that the lint should be on "non-thin pointers" where https://github.com/rust-lang/rust/issues/106447#issuecomment-1769077122 says specifically for "dyn trait pointer comparison".

I believe that the *intent* was only to lint on dyn trait, because it could be useful to compare the metadata portion of slice raw pointers. This does appear to be what is implemented, but I think it would be good to get clarification on exactly what was agreed on.

**Comment by asquared31415** [comment]
Naming bikeshed since you mentioned that "invalid" seems strong: `unreliable_dyn_pointer_comparisons`.

I think the name should communicate that the comparison may spuriously return `true` or `false` and cannot be relied upon for correctness.

**Comment by Urgau** [comment]
@scottmcm could you clarify https://github.com/rust-lang/rust/pull/117758#issuecomment-1804319331 ?

As far as naming goes, it will depend on an answer to the previous question.

**Comment by asquared31415** [comment]
The [Reference](https://doc.rust-lang.org/beta/reference/types/pointer.html#raw-pointers-const-and-mut) says
> When comparing raw pointers they are compared by their address, rather than by what they point to. When comparing raw pointers to [dynamically sized types](https://doc.rust-lang.org/beta/reference/dynamically-sized-types.html) they also have their additional data compared.

The following code also executes without panicking:
```rust
fn main() {
    let arr = [1_u32, 2, 3, 4];
    let ptr = arr.as_ptr();
    let p1 = core::ptr::slice_from_raw_parts(ptr, 4);
    let p2 = core::ptr::slice_from_raw_parts(ptr, 3);
    // addresses are equal
    assert!(p1.cast::<()>() as usize == p2.cast::<()>() as usize);
    // pointers are not equal
    assert!(p1 != p2);
}
```

I think that it's reasonable to rely on the metadata being compared.  The issue arises because `*const dyn Trait` metadata isn't consistent, so when the metadata *is* compared, the comparison might be meaningless for such types.

**Comment by asquared31415** [comment]
Also there's not a stable way to get the length metadata of a slice pointer without asserting validity as a reference to 0 bytes (you can make a `&[()]` and access its length), so I doubt that many users are using that method that you linked at all.

File: compiler/rustc_lint/src/types.rs [hunk 1]
```diff
@@ -136,6 +137,37 @@ declare_lint! {
...
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The comparison include metadata which is not guaranteed to be equivalent.
+    INVALID_POINTER_TRAIT_COMPARISONS,
+    Warn,
+    "detects invalid pointer trait comparisons"
```
Comment (RalfJung) [inline compiler/rustc_lint/src/types.rs:168]: "invalid" is a strong word to use here. The comparisons aren't really invalid, are they? They just behave in surprising ways.

**Comment by Urgau** [comment]
> Hmm, I think my personal vision of this was something like `unclear_fat_pointer_comparisons` that would lint on `*const T` for any `T: ?Sized`

In the light of this, I've adjusted the lint accordingly and renamed it as you suggested.

**Comment by joshtriplett** [comment]
Request: could we call this `ambiguous_dyn_pointer_comparison`, since the issue is primarily about `dyn` pointers and the ambiguity of including or excluding the vtable? We could name `ambiguous_slice_pointer_comparison` separately.

**Comment by traviscross** [comment]
We discussed this in the T-lang meeting today.  The consensus was to do something along the lines of what @scottmcm proposed.  To improve the recommendation we can make to users, T-libs-api proposed to stabilize `ptr::addr_eq`, and that's now in FCP in #116324.  On this issue @scottmcm will be making a concrete proposal for FCP merge.

**Review by davidtwco — APPROVED** [review]
Implementation LGTM, r=me after various FCPs, renamings, etc are decided and resolved.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

As discussed above, we're waiting on a concrete proposal here, so we'll go ahead and remove the nomination for now.  Please renominate when that proposal is made.

**Comment by joshtriplett** [comment]
Two bikesheds:

- Could we change "unclear" to "ambiguous"? I think that's more specific: it's *ambiguous* whether you want to include or exclude the metadata (for dyn) or the length (for slice), so you should use a method that makes that explicit.
- Could we please have two separate lint gates, one for `ambiguous_dyn_comparison` and another for `ambiguous_slice_ptr_comparison`? There are, as far as I can tell, two kinds of pointers this applies to, slice pointers and dyn pointers, so rather than pushing a jargon term on folks in the lint name, let's actually describe the two cases.

@rfcbot concern naming
@rfcbot reviewed

**Comment by joshtriplett** [comment]
@scottmcm I edited my comment in a couple of ways, and in particular withdrew the suggestion to have two lints.

**Comment by Urgau** [comment]
@scottmcm I've update the PR to lint on `==`, `!=` and `PartialEq::{eq,ne}` on wide pointers (module references) and have the diagnostics suggest `ptr::eq` and `ptr::addr_eq` appropriately.

@joshtriplett As asked I've used the "wide" terminology and renaming the lint to `ambiguous_wide_pointer_comparisons`. This should hopefully resolve your naming concern.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Let's renominate as it seems we need to settle the naming question.  In particular, if we want to call these "wide pointers" rather than ["fat pointers"](https://en.wikipedia.org/wiki/Dynamic_dispatch#Fat_pointer), then we probably need to be sure we're also happy with whatever the antonym should be, as it might seem inconsistent to have "wide" and "thin" pointers.  We have not previously used "narrow pointers" as a term of art.

In terms of the resources we might want to update if we were to change our terminology:

- Wikipedia uses [fat pointer](https://en.wikipedia.org/wiki/Dynamic_dispatch#Fat_pointer) as a term of art for Rust and some other languages.  We would want to post a canonical resource on our site that Wikipedia editors could cite to distinguish Rust as having "wide pointers" rather than the fat ones in other languages.
- There's a StackOverflow [post](https://stackoverflow.com/questions/57754901/what-is-a-fat-pointer#57754902) that's highly ranked by search engines describing what "fat pointers" are in Rust.
- The `rustc` dev guide [places](https://rustc-dev-guide.rust-lang.org/appendix/glossary.html) the relevant content under "fat pointer", but notes that fat pointers may also be known as "wide pointers" or "double pointers".
- These RFCs all refer to fat pointers:
  -  [RFC 198][]
  - [RFC 401][]
  - [RFC 1135][]
  - [RFC 1211][]
  - [RFC 1845][]
  - [RFC 1861][]
  - [RFC 2580][]
- These RFCs refer to "fat arrows".  Might we want to change this terminology also if we choose to avoid fat pointers?
  - [RFC 90][]
  - [RFC 550][]
  - [RFC 2535][]

By way of contrast, [RFC 3348][] and [RFC 3324][] refer to "wide pointers".

[RFC 90]: https://github.com/rust-lang/rfcs/blob/master/text/0090-lexical-syntax-simplification.md
[RFC 198]: https://github.com/rust-lang/rfcs/blob/master/text/0198-slice-notation.md
[RFC 401]: https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md
[RFC 550]: https://github.com/rust-lang/rfcs/blob/master/text/0550-macro-future-proofing.md
[RFC 1135]: https://github.com/rust-lang/rfcs/blob/master/text/1135-raw-pointer-comparisons.md
[RFC 1211]: https://github.com/rust-lang/rfcs/blob/master/text/1211-mir.md
[RFC 1845]: https://github.com/rust-lang/rfcs/blob/master/text/1845-shared-from-slice.md
[RFC 1861]: https://github.com/rust-lang/rfcs/blob/master/text/1861-extern-types.md
[RFC 2535]: https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md
[RFC 2580]: https://github.com/rust-lang/rfcs/blob/master/text/2580-ptr-meta.md
[RFC 3324]: https://github.com/rust-lang/rfcs/blob/master/text/3324-dyn-upcasting.md
[RFC 3348]: https://github.com/rust-lang/rfcs/blob/master/text/3348-c-str-literal.md

**Comment by LunarLambda** [comment]
I would be in favour of wide/narrow (or wide/thin, it doesn't feel too unnatural/mismatched to me personally).

But also, I find the allusions to the pointer's size a bit awkward, since that's not the primary thing people care about. For example, Near/Far pointers are different sizes too, but that's secondary to their actual capabilities: Being able to refer to things inside or outside the current segment (in the case of far pointers).

Maybe we could find a pair of terms that addresses the capability of "wide" pointers to carry metadata rather than their size? (since that's the current language for [Pointee](https://doc.rust-lang.org/core/ptr/trait.Pointee.html) and other parts of std)

Although admittedly, as a non-native speaker I don't have an immediate suggestion for suitable pairs. "extended" comes to mind, but is a lot longer than "wide".

**Comment by joshtriplett** [comment]
I'm resolving my concern on the basis of the rename to `wide`.

@rfcbot resolved naming

We can, separately, consider if we want to determine a single unified term for systematic use. But that doesn't need to be a blocker here.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

This was discussed in triage, and the consensus is that this is OK to go forward with the name change.  Let's remove the nomination.

**Comment by m-ou-se** [comment]
It might make sense to suggest both `eq` and `addr_eq` for slice pointers, but *only* suggest `addr_eq` (and not `eq`) for trait object pointers. (Because vtable addresses aren't unique/reliable.)

**Comment by m-ou-se** [comment]
Other than that, no concerns came up in the libs-api meeting. :) Thanks for working on this!

**Comment by Urgau** [comment]
@davidtwco Since you reviewed this PR, the scope of the lint grew a bit to lint on all unsized pointers as such the implementation changed quite a bit, in particular the second commit https://github.com/rust-lang/rust/pull/117758/commits/5630540ac0536558c929cce0f1db5e9b8cfd0582.

I would appreciate it if you could review those changes or delegate me bors permissions for this PR.

**Review by davidtwco — APPROVED** [review]

**Comment by RalfJung** [comment]
> Testing GCC stage1

@GuillaumeGomez @antoyo any idea why this would land but then start failing?

**Comment by GuillaumeGomez** [comment]
@Urgau proposed a theory about this [here](https://github.com/rust-lang/rust/pull/118835#issuecomment-1850612255). I'm still surprised though.

**Pull request merged**